### PR TITLE
Fix shared pointers initialization in system_parameters and cpu_info classes

### PR DESCRIPTION
**Pull request by Vika-F**
## Description

The initial idea was to make pointers to implementation `impl_` unique pointers in both `system_parameters` and `cpu_info` classes.
But it resulted in the need of similar changes in the derived classes.

So, the `impl_` pointers types were left unchanged, but their initialization was changed to `new class_name()`.

Still the use of shared pointers (instead of unique) as pointers to implementations looks as not the best design choice.

---

PR should start as a draft, then move to ready for review state after CI is passed and all applicable checkboxes are closed.
This approach ensures that reviewers don't spend extra time asking for regular requirements.

You can remove a checkbox as not applicable only if it doesn't relate to this PR in any way.
For example, PR with docs update doesn't require checkboxes for performance while PR with any change in actual code should have checkboxes and justify how this code change is expected to affect performance (or justification should be self-evident).

Checklist to comply with **before moving PR from draft**:

**PR completeness and readability**

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/uxlfoundation/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_.
- [x] I have added a respective label(s) to PR if I have a permission for that.
- [x] I have resolved any merge conflicts that might occur with the base branch.

**Testing**

- [x] I have run it locally and tested the changes extensively.
- [x] All CI jobs are green or I have provided justification why they aren't.

The failures in distributed tests do not look related to the changes.

**Performance**

not applicable

### CODE REVIEW

**Comment by Vika-F** [comment]
@david-cortes-intel 
> Thanks, this looks like an improvement, but there might still be something wrong in this same class even after all the changes here - see logs:

Thank you for the diagnostic. For now I do not understand why there is a null pointer as a result of make_shared. I will try to figure it out.

**Comment by david-cortes-intel** [comment]
> @david-cortes-intel
> 
> > Thanks, this looks like an improvement, but there might still be something wrong in this same class even after all the changes here - see logs:
> 
> Thank you for the diagnostic. For now I do not understand why there is a null pointer as a result of make_shared. I will try to figure it out.

Perhaps something to do with casts done in class `base`?

File: cpp/oneapi/dal/detail/cpu_info.cpp [hunk 1]
```diff
@@ -29,24 +29,28 @@
 namespace oneapi::dal::detail {
 namespace v1 {
 
-cpu_info::cpu_info() {
+cpu_info::cpu_info()
+        :
 #if defined(TARGET_X86_64)
-    impl_ = detail::pimpl<cpu_info_iface>(std::make_unique<cpu_info_x86>());
+          impl_(new cpu_info_x86)
```
Comment (david-cortes-intel) [inline cpp/oneapi/dal/detail/cpu_info.cpp]: Would this type of call without parentheses leave the memory uninitialized?

Also from internal discussions it seems that it shouldn't be a problem to break ABI compatibility only around internal-only functions.

Comment (Vika-F) [reply cpp/oneapi/dal/detail/cpu_info.cpp]: In C++ there is no difference between `new T` and `new T()` in case T's default constructor is defined. Anyway I've added the `()` to make the code more understandable.

> Also from internal discussions it seems that it shouldn't be a problem to break ABI compatibility only around internal-only functions.

Just made the ABI checker happier.
Also, as you pointed before, the use of `new` is more compliant with the rest of the codebase.

**Comment by david-cortes-intel** [comment]
@Vika-F I managed to run UBSan in sklearnex now, got this new info which looks very relevant:
```
cpp/daal/src/services/service_topo.cpp:780:45: runtime error: left shift of negative value -1
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior cpp/daal/src/services/service_topo.cpp:780:45 
cpp/daal/src/services/service_topo.cpp:786:50: runtime error: left shift of negative value -1
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior cpp/daal/src/services/service_topo.cpp:786:50
```
